### PR TITLE
Refine playback

### PR DESCRIPTION
**Pull request by stevenpetryk**
This adds some enhancements to playback

- [x] Added ability to play and pause streams
- [x] Added ability to skip to random points in streams
- [x] Gracefully handle when a stream ends
- [x] Add proof-of-concept audio playback (with Khan Academy audio)
- [x] Optimized skipping forward (just draws line-by-line, not point-by-point)
- [x] Added a loading state to the stream view

Greg, I'm putting you in this PR, because you told me you haven't built the frontend in 2 months. ;)

### CODE REVIEW

File: src/actions/stream.js [hunk 1]
```diff
@@ -34,26 +34,35 @@ export function fetchStreamData (streamId) {
...
-              })
-            }
-          },
-          'STREAM_DATA_FAILURE']
+    return new Promise((resolve) => {
+      if (!endpoint) {
+        setTimeout(() => {
+          dispatch(fetchStream(streamId))
+          dispatch(fetchStreamData(streamId)).then(resolve)
+        }, 1000)
```
Comment (stevenpetryk) [inline src/actions/stream.js:42]: This just makes the frontend retry until the upload to S3 is complete. In testing, it rarely retries more than once.

File: src/components/playback/DrawManager.js [hunk 2]
```diff
@@ -0,0 +1,211 @@
...
+
+    const initialPosition = this.position
+
+    const frame = (now) => {
+      if (!start) start = now
+
+      this.position = clamp(initialPosition + now - start, 0, this.duration)
+      this.emit('POSITION_CHANGE', this.position)
+
+      if (this._draw()) {
```
Comment (stevenpetryk) [inline src/components/playback/DrawManager.js:86]: `_draw()` returns false when there's nothing left to draw (i.e., once the stream is done playing).

File: src/components/playback/DrawManager.js [hunk 3]
```diff
@@ -0,0 +1,211 @@
...
+    return true
+  }
+
+  _getParsedLine (index) {
+    if (!this.parsedLines[index]) {
+      this.parsedLines[index] = JSON.parse(this.lines[index] || 'null')
+    }
+
+    return this.parsedLines[index]
+  }
```
Comment (stevenpetryk) [inline src/components/playback/DrawManager.js:152]: Lazy parsing ftw!

File: src/actions/stream.js (hunk 1 above)
Comment (maxcell) [reply src/actions/stream.js:42]: I think that should be fine! As long as it really isn't a noticeable thing. What about if the upload fails for whatever reason? Should we worry about that?

File: src/components/playback/DrawManager.js [hunk 4]
```diff
@@ -0,0 +1,211 @@
...
+    this.playing ? this.pause() : this.play()
+  }
+
+  stop () {
+    allOff(this)
+  }
+
+  setPosition (position) {
+    const wasPlaying = this.playing
+    if (this.playing) this.pause()
```
Comment (maxcell) [inline src/components/playback/DrawManager.js]: Should this be `wasPlaying`?

Comment (maxcell) [reply src/components/playback/DrawManager.js]: `if (wasPlaying) this.pause()`?

File: src/components/playback/DrawManager.js [hunk 5]
```diff
@@ -0,0 +1,211 @@
...
+    this.lineCursor = 0
+    this.pointCursor = 0
+    this.currentLine = null
+    this.path = null
+  }
+}
+
+EventEmitter(DrawManager.prototype)
+
+/* global d3, requestAnimationFrame, cancelAnimationFrame */
```
Comment (maxcell) [inline src/components/playback/DrawManager.js:223]: What's going on here?

File: src/actions/stream.js (hunk 1 above)
Comment (stevenpetryk) [reply src/actions/stream.js:42]: We should worry about that, but I don't think the frontend should.

File: src/components/playback/DrawManager.js (hunk 4 above)
Comment (stevenpetryk) [reply src/components/playback/DrawManager.js]: That works too! Probably more descriptive. Thanks.

File: src/components/playback/DrawManager.js (hunk 5 above)
Comment (stevenpetryk) [reply src/components/playback/DrawManager.js:223]: These comments tell the linter "ay, these are defined in the global scope, so don't get mad at me for referencing undefined variables"

Comment (maxcell) [inline src/components/playback/DrawManager.js]: 👍

**Review by maxcell — APPROVED** [review]
Personally love it! It would be neat to have playback reverse/forward by 5 sec using the arrows but super low priority and use case.

**Pull request merged**